### PR TITLE
Fix failing futures from ChapelArray -> ChapelDomain separation.

### PR DESCRIPTION
**Pull request by stonea**
Fix `.bad` files for failing futures caused by this PR, which separates `_domain` and associated functions out into their own file: https://github.com/chapel-lang/chapel/pull/18829.

To reviewers ---

* I don't think there's anything controversial in updating `test/distributions/ferguson/dist-assoc-bulkadd.bad`, just updating filenames and line numbers for code I moved
* For the most part the same is true for `test/classes/generic/typeOfGenericField-fromTypeVariable.bad`. One thing that looks odd to me is it looks like the error for:

`unresolved call 'type [BlockDom(1,int(64),false,unmanaged DefaultDist)] int(64)._instance'`

Gives as a candidate function:
`$CHPL_HOME/modules/internal/ExternalArray.chpl:32: note:   chpl_opaque_array._instance`

Instead of:

`$CHPL_HOME/modules/internal/ChapelArray.chpl:1187: note:   _domain._instance`

Kind of confusing since I didn't touch `ExternalArray.chpl` at all but I think maybe reordering the functions has impacted name resolution?

I'm tempted to say I should just update the .bad file and move on (especially since this is locking down .bad behavior), but if this raises a red flag for anyone let me know and I can dig deeper.

* I'm also confused why I didn't see these failures in paratests. I'm can see in my log file they didn't run but I don't know any reason why not. I don't see any `.skipif` file or any obvious reason it wouldn't run. Do you know if you need to pass some special flag to paratests or something to indicate to run futures?

### CODE REVIEW

**Comment by lydia-duncan** [comment]
Yeah, you need to run with `-futures` in your paratest runs

**Review by lydia-duncan — APPROVED** [review]
I agree, these changes don't worry me - I think you're right that adding a new module changed the order in which resolved candidates were listed and that since it was already failing, that isn't anything to worry about (especially since the number of candidates for the failure didn't change)

**Comment by stonea** [comment]
> Yeah, you need to run with -futures in your paratest runs

Ok that explains it; thanks. I might update my tool to make that the default behavior. Do you know if there was any particular reason it wasn't made the default behavior for paratests? Seems like we would always want to know if we broke a .bad file if it's going to cause failures in Jenkins.

**Comment by lydia-duncan** [comment]
I suspect it's for faster iteration when testing changes to a branch.  Personally, I always run with `-futures`, but I can see wanting to get a sense of failure cases without wanting the clutter that can come from the output of futures.  It also means you can search for `[Error` in the log file and not run into a failing future whose behavior didn't change as a result of your modifications

**Comment by bradcray** [comment]
I think Lydia's explanation is correct.  If I have a PR that is in the early stages, I test without futures to see how bad it is quickly.  Once I get closer to being ready to merge (or if I'm in no rush to get the results), I run with `-futures`.  From a testing noise perspective, I agree that having the tool default to testing futures is probably safer; if there were a way to opt out for people who wanted to go faster, that could be nice.

**Comment by stonea** [comment]
> I think Lydia's explanation is correct. If I have a PR that is in the early stages, I test without futures to see how bad it is quickly. Once I get closer to being ready to merge (or if I'm in no rush to get the results), I run with -futures. From a testing noise perspective, I agree that having the tool default to testing futures is probably safer; if there were a way to opt out for people who wanted to go faster, that could be nice.

Ok now I'm curious how much faster it is without testing futures. I'll have to run that experiment sometime.

Seems like we could change the default behavior to `paratests` to be run all futures unless you pass a `--fast`
flag.  Maybe I'll add this to my list of "nice to do someday projects" (I'd say "rainy day" project but
that's like 90% of the days now until March in Portland :-)).

Although I guess strictly as a user of `chapcs` my preference would be that everyone else get their jobs finished as quickly as possible; although if skipping futures by default causes people to have to rerun paratests multiple times maybe it isn't any more efficient from that perspective?  There's probably some analogy you could make about low-flow toilets not always saving water there but I won't :-P.

Maybe the best thing to do would be to have some sort of staged job. Do all the non-futures first and if there's any failures stop and report back to the user; if everything looks good then do another pass that's a "futures" only job.

**Comment by lydia-duncan** [comment]
Yeah, that's reasonable!  There is already a `-futuresonly` mode for paratests

**Comment by bradcray** [comment]
> Seems like we could change the default behavior to paratests to be run all futures unless you pass a --fast
flag.

If we were to add this, I might call it `--asap` to avoid any implications that this means "compile passing `--fast` to `chpl`".  That said, since this is for developers, I might just teach them to throw `--no-futures` unless we come up with other things that `--asap` (or whatever) would control.  Otherwise, it just raises the question "What's the difference between these?"

**Comment by stonea** [comment]
> If we were to add this, I might call it --asap to avoid any implications that this means "compile passing --fast to chpl". That said, since this is for developers, I might just teach them to throw --no-futures unless we come up with other things that --asap (or whatever) would control. Otherwise, it just raises the question "What's the difference between these?"

I was going to suggest `--no-futures` at first but anyone wanting to speed-up their testing time isn't going to want to wait around typing 7 extra characters (compared to `--fast`). `--asap` works too though I agree (like `--fast`) it's less specific about exactly what we're doing to speed testing up.

**Pull request merged**
